### DISCUSSION TITLE
Add abstraction for Theme (dark/light) switching

### DISCUSSION THREAD

**Discussion by nickrandolph** [unoplatform/uno.extensions#420] [Ideas]
Provide an abstraction where the theme (light or dark mode) of the app can be switched (and persisted)

**Comment by francoistanguay** [comment]
If it can be persisted, I think we also need a mode that uses the "System" instead of a Light/Dark override... More like Default/Light/Dark where Default could clear the stored override or something

**Reply by francoistanguay** [reply]
Unclear how Default makes sense as a Value at the public api level though. It's mostly something internal. Although if you override, you want to be able to clear the override, so we need to come up with some form of API for it all.

**Comment by arqueror** [comment]
Proposed interface:

```
	public interface IThemeService
	{
	        bool UseSystemTheme {get; set;}
		bool IsDarkMode{ get; }
		Task SetThemeAsync(bool darkMode);
		Task SetThemeAsync();
	}
```

**Reply by francoistanguay** [reply]
I feel we need something like:
DesiredTheme ^System, Light, Dark } { get; set; }  // or get + SetDesiredThemeAsync
CurrentTheme { Light, Dark } { get; }
CurrentThemeCHanged

DesiredTheme is for those cases where a user wants to override the system default (or rely on system) In some apps, it's represented by a Radiobutton:

![image](https://user-images.githubusercontent.com/30851132/203659919-44483001-e585-43d5-a01a-1f1b90bf56e5.png)

WHereas CurrentTheme might be useful if you need to react dynamically to the Current Theme changing...

**Reply by francoistanguay** [reply]
And as it's already scoped by the IThemeService, properties could be named simply DesiredValue and CurrentValue maybe
@arqueror  @nickrandolph

**Reply by nickrandolph** [reply]
I agree, just simplify a bit
```
DesiredTheme { System, Light, Dark}
SetThemeAsync(DesiredTheme);
DesiredThem Theme {get;}
ThemeChanged;
```

Also, for first iteration ThemeChanged would just be raised when the theme is changed by the ThemeService. However, ideally we would want this to capture (where possible the system theme changed too).

**Reply by francoistanguay** [reply]
I think we need 2 gets... The desired one, and the effective one that is either Light or Dark.... First is for the setting itself that can read/write, the effective one is to expose the outcome of the desire., could be used in app behaviors/adaptive stuff that wants to react to theme change....And when reactive, you cant have System as an effective value, just the outcome being LIght or Dark.

**Reply by francoistanguay** [reply]
So we still need DesiredTheme {System, Light, Dark} and could be just IsDark : bool if we dont want a Light/Dark enum

**Reply by nickrandolph** [reply]
Ok makes sense, so

```
DesiredTheme { System, Light, Dark}
SetDesiredThemeAsync(DesiredTheme);
DesiredThem DesiredTheme {get;}
DesiredThemeChanged;

bool IsDark {get; }
ThemeChanged?
```
Do we need the second event, or do we just have a single event that gets raised if either (or both) desired and/or actual theme has changed?

**Reply by francoistanguay** [reply]
Well, to be consistent, the IsDark even would be IsDarkChanged probably. We can add it later if better.

**Reply by arqueror** [reply]
Updated interface proposal:
```
	public interface IThemeService
	{
		bool IsDark { get; }
		DesiredTheme Theme { get; }
		Task SetThemeAsync(DesiredTheme theme = DesiredTheme.System);
		event EventHandler<DesiredTheme> DesiredThemeChanged;
	}
```

**Reply by dansiegel** [reply]
Perhaps I've missed something here, but this issue seems to be missing a the context of the SDK we're already working with. We already have:
```cs
public enum ApplicationTheme
{
    Light,
    Dark
}

public enum ElementTheme
{
    Default,
    Light,
    Dark
}
```

Why exactly do we need to introduce yet another enum that replicates what we already have to work with? This whole thing seems like it can be simplified to:

```cs
public interface IThemeService
{
    // Use ApplicationTheme here because it will either be
    // Light or Dark
    ApplicationTheme CurrentTheme { get; }

    // Use ElementTheme here because we may want to set Default
    // to revert to the OS preference for Light/Dark
    Task SetThemeAsync(ElementTheme theme);

    event EventHandler<ApplicationTheme> ThemeChanged;
}
```

I don't really see a reason why we would want to have a boolean for IsDark when we have an Enum that has 2 choices

**Reply by francoistanguay** [reply]
First, I think they have different meanings .... Second, they are in a WinUI-defined assembly and IThemeService is meant to be used in a presentation layer that is UI-tech agnostic...

You could argue that we could easily cast one into the other where applicable, but am unclear what we're using in our SystemThemeHelper under the hood...

**Reply by francoistanguay** [reply]
But I'd have no issue converting Light/Dark into an enum in Uno Extensions instead of beingn a boolean if it makes it more symmetric/cleaner

**Reply by dansiegel** [reply]
Yeah but it should still be built against either WinUI/UWP which both of them have those enums so we shouldn't have an issue using them.
